astropy/nddata/tests/test_nddata.py

```diff
@@ -587,3 +594,99 @@ def test_overriden_wcs():
     # Check that a sub-class that overrides `.wcs` without providing a setter
     # works
     NDDataCustomWCS(np.ones((5, 5)))
+
+
+# set up parameters for test_collapse:
+np.random.seed(42)
+collapse_units = [None, u.Jy]
+collapse_propagate = [True, False]
+collapse_data_shapes = [
+    # 3D example:
+    # (4, 3, 2),
+    # 5D example
+    (6, 5, 4, 3, 2)
+]
+collapse_masks = list(
+    chain.from_iterable(
+        [
+            # try the operations without a mask (all False):
+            np.zeros(collapse_data_shape).astype(bool)
+        ]
+        + [
+            # assemble a bunch of random masks:
+            np.random.randint(0, 2, size=collapse_data_shape).astype(bool)
+            for _ in range(10)
+        ]
+        for collapse_data_shape in collapse_data_shapes
+    )
+)
+permute = len(collapse_masks) * len(collapse_propagate) * len(collapse_units)
+collapse_units = permute // len(collapse_units) * collapse_units
+collapse_propagate = permute // len(collapse_propagate) * collapse_propagate
+collapse_masks = permute // len(collapse_masks) * collapse_masks
```

Comment:
Can you add a comment somewhere in the block above explaining the intent these constructions? I don't see a problem with the code though my head is spinning a bit digesting it. A short guide to it would be helpful

Reply:
Absolutely. It's a tricky passage. Opaque enough that it's maybe slightly evil 😈